analysis/known_objects.py

```diff
@@ -0,0 +1,361 @@
+            sc : a SkyCoordinate object 
+                Indicates the position of the object.
+        """
+        if self.max_time_index < time_index:
+            self.max_time_index = time_index
+
+        # If we haven't seen this object before add an array
+        # to store the observations.
+        if name not in self.objects:
+            self.objects[name] = []
+
+        # Fill in any missing observations with None.
+        while len(self.objects[name]) < time_index - 1:
+            self.objects[name].append(None)
+
+        # Add the new observation.
+        self.objects[name].append(sc)
+
+    def get_num_times_seen(self, name):
+        """
+        Count the number of times we saw an object.
+        
+        Arguments:
+            name : string
+                Object name from the databse.
+                
+        Returns:
+            integer : the count of times we should have
+                      seen this object.
+        """
+        if not name in self.objects:
+            return 0
+
+        count = 0
+        for sc in self.objects[name]:
+            if sc is not None:
+                count = count + 1
+        return count
+
+    def add_observation(self, name, time_index, sc):
+        """
+        Add an observation to the set seen known objects.
+        
+        Arguments:
+            name : string
+                Object name from the databse.
+            time_index : integer
+                The index of the timestep for this observation.
+            sc : a SkyCoordinate object 
+                Indicates the position of the object.
+        """
+        if self.max_time_index < time_index:
+            self.max_time_index = time_index
+
+        # If we haven't seen this object before add an array
+        # to store the observations.
+        if name not in self.objects:
+            self.objects[name] = []
+        arr = self.objects[name]
+
+        # Fill in any missing observations with None.
+        while len(arr) < time_index - 1:
+            arr.append(None)
+
+        # Add the new observation if there isn't one already.
+        if len(arr) <= time_index:
+            arr.append(sc)
+
+    def pad_results(self):
+        """
+        Pad the result vectors with None so that they
+        are all the same length.
+        """
+        for name in self.objects.keys():
+            # Fill in any missing observations with None.
+            while len(self.objects[name]) < self.max_time_index:
+                self.objects[name].append(None)
+
+    def filter_observations(self, num_obs):
+        """
+        Remove all objects from the result list that appear
+        fewer than the given number of times.
+
+        Arguments:
+            num_obs - The minimum number of images the known object
+                      must appear in to be counted.
+        """
+        filtered_results = {}
+        for name in self.objects.keys():
```

Comment:
This is stealthily O(n^2) where n=objects.size. Is n going to get potentially large?

Reply:
I rewrote the algorithm.